README.md
<!--
 @since 2023.02.09, 16:49
 @changed 2023.02.13, 01:28
-->

# Tournament game backend server


## Build info (auto-generated)

- Version: 0.0.1
- Last changes timestamp: 2023.02.13, 01:15
- Last changes timetag: 230213-0115


## TODO

Describe used techologies & their relations.


## Install

Install all required node dependencies:

```
npm i
```

Initialize python virtual environment:

```
sh utils/venv-init.sh
```


## Start dev server

Acivate virtual environment:

On windows:

```
call .venv/Scripts/activate
```

On unix:

```
. ./.venv/Scripts/activate
```

Start local flask server (can be located in browser with `http://localhost:5000/*`, but mostly using for api requests):

```
npm run start
```

Or just:

```
flask run
```


## API

Basic api structure:

TODO: Describe api here (for specific server version)


## Server urls

Remote server: `https://tournament-game.march.team/`
Remote server: `https://tournament-game-build.march.team/` (with built app)


## Server

Server runs on python/flask platform.

TODO: Describe basic server functionality.


## Python venv maintenance

Server command for creating venv:

```
virtualenv -p python3 ~/.venv-py3-flask
source ~/.venv-py3-flask/bin/activate
pip install -r requirements.txt
```

Local script for venv creating and initialization:

```
sh utils/util-venv-init.sh
```

Local command for activate venv:

```
call .venv/Scripts/activate
source .venv/Scripts/activate
```


## Python dependencies

```
pip install PKGNAME
pip install -r requirements-general.txt -r requirements-dev-only.txt
pip freeze > requirements-frozen.txt
```

Use `utils/venv-init.*` scripts.



CHANGELOG.md
<!--
@changed 2022.04.03, 21:46
-->

# Changelog

- 2022.04.03, 21:46 -- Mail sending minor changes: added `legalOrigins` hosts for `localhost` only in dev mode, updated mail message styles, tried to add links for favicons (in message body).
- 2022.04.02, 18:26 -- Mail message: added formatting, content preview (as plain body), footer.
- 2022.04.02, 17:55 -- Excplicitly set encoding as `utf-8` for config helper methods `readFiletoString` and `updateConfigWithYaml`. Specified parameters `encoding='UTF-8'`` and `ensure_ascii=False` for json datafile opening in `RequestsStorage`.
- 2022.04.02, 17:13 -- Send mail using `smtp.fullspace.ru` server and account `site@march.team`.
- 2022.04.02, 16:16 -- Added more sophisticated session processng. Session tokens checks at start with checkToken=False flag. Other api methods must process verification with checkToken=True (default) value.
- 2022.04.02, 16:13 -- Used march.team.realty google account for mail sending. Disabled actual mail sending for dev server.
- 2022.03.29, 23:23 -- Added request origin checks for `blueprintRequests` module api.
- 2022.03.29, 23:20 -- Use `Token` instead `sessionId`. Added api methods verificartion for `origin` header and compare it for `config.legalOrigins`. Removed static app serve functinality (slow and buggy).
- 2022.03.28, 15:35 -- Using production build.
- 2022.03.28, 15:32 -- v.0.0.2: Renamed project from `march-team-back` to `march-team-flask-site`.
- 2022.03.28, 00:03 -- Moved mail address constants to config, added dev-mode mail sending only to bcc addresses.
- 2022.03.27, 23:43 -- Advanced errors processing (returns such request data as url, protocol, method). check request data for `requests/add` request.
- 2022.03.27, 21:37 -- Moved core server code to `blueprintCoreSite` and `app` modules. Support for static spa app build pages.
- 2022.03.27, 19:38 -- Added `html-app*` folder support for spa app bundle.
- 2022.03.26, 04:47 -- Using CORS (Flask-Cors).
- 2022.03.26, 02:28 -- Minor changes (config, auto timestamps in RequestsStorage, api password).
- 2022.03.26, 02:08 -- Configuration for google mail server and mail message styles.
- 2022.03.26, 01:16 -- Added mail sending functionality.
- 2022.03.25, 23:05 -- Minimal api (root, records), requests database (using tinydb).
- 2022.02.27, 02:45 -- RecordsStorage using TinyDB (fully worked code & tests).
- 2022.02.26, 05:31 -- Using TinyDB for data engine in `RecordsStorage` (in progress: basic methods and tests are ready).
- 2022.02.24, 00:55 -- v.0.0.5: Moved `lib` to `core/lib` scope (extract to shared submodule in future). Using pylint as project-wide linter (fixed all linter issues).
- 2022.02.23, 00:24 -- RecordsStorage: remove outdated records during findng (in `processRecords`), added metod for rorced remove of outdated records (`removeOutdatedRecords`).
- 2022.02.22, 02:30 -- RecordsStorage, Record: Basic records and records storage routines and tests.
- 2022.02.21, 23:06 -- Test support utilities (`getTrace`). Restuctured utils module.
- 2022.02.15, 06:12 -- v.0.0.4: Created automatic tests & linters for python code.
- 2022.02.15, 03:53 -- Helper modules moved to lib folder.
- 2022.02.14, 04:45 -- Experimental sessions support.
- 2022.02.12, 04:52 -- Implemented simple socket.io tests. Relocated cdn files.
- 2022.02.12, 02:58 -- v.0.0.3: Restructured server application (extracted several blueprint modules for different apis).
- 2022.02.08, 06:19 -- Device server shot creation method.
- 2022.02.08, 01:51 -- Fixed logger async (non-atomic) print issuesm server: starting only once (skip first initialization in debug mode, when app is initialized twice), templates: temporarily removed unused assets.
- 2022.02.07, 22:33 -- Local device server start scripts (via gunicorn, see `utils/rpi-start-server.sh`).
- 2022.02.07, 03:24 -- Readme venv instructions, disable basic auth in htaccess.


config.py
# -*- coding:utf-8 -*-
# @module config
# @desc Universal server & client config
# @since 2022.02.06, 23:56
# @changed 2023.02.09, 22:39
# See:
#  - https://docs.python.org/3/library/configparser.html -- ???
#  - https://stackoverflow.com/questions/9590382/forcing-python-json-module-to-work-with-ascii
#  - https://flask.palletsprojects.com/en/2.0.x/config/

import os
from os import path
import json
#  import yaml
import sys

from config_helpers import updateConfigWithYaml, readFiletoString

pythonVersion = sys.version

#  Determine running under test suite...
isTest = 'unittest' in sys.modules

gunicornEnv = os.getenv('GUNICORN_ENV')  # Detect gunicorn server (run on a raspberry pi)
flaskEnv = os.getenv('FLASK_ENV')  # detect flask server (dev mode)
isGunicorn = bool(gunicornEnv)
isRPi = isGunicorn
isDev = flaskEnv == 'development'
isProd = not isDev
# TODO: Detect test environment

rootPath = path.dirname(path.abspath(__file__))  # Project root path

yamlConfigFilename = path.join(rootPath, 'config.yml')
yamlLocalConfigFilename = path.join(rootPath, 'config.local.yml')

uploadPath = path.join(rootPath, 'uploads')

# Project structure

# (UNUSED) See link creation in `utils/make-html-app-links.sh`, `utils/make-html-app-links-local.sh`
#  Generate/read build parameters (version, timetag etc)
#  Default values (empty)...
version = ''
timestamp = ''
timetag = ''
buildTag = ''
#  Filenames...
buildVersionFilename = path.join(rootPath, 'build-version.txt')
buildTagFilename = path.join(rootPath, 'build-tag.txt')
timestampFilename = path.join(rootPath, 'build-timestamp.txt')
timetagFilename = path.join(rootPath, 'build-timetag.txt')
packageFilename = path.join(rootPath, 'package.json')
#  Read version...
#  print('config: packageFilename', packageFilename  # DEBUG)
if path.isfile(buildVersionFilename):
    version = readFiletoString(buildVersionFilename, 'UNSPECIFIED')
elif path.isfile(packageFilename):
    with open(packageFilename, encoding='utf-8') as pkgConfigFile:
        pkgConfig = json.load(pkgConfigFile)
        version = pkgConfig['version'].encode('ascii')
        #  pkgConfigFile.close()
# Read timestamp/timetag...
if path.isfile(timestampFilename):
    timestamp = readFiletoString(timestampFilename)
if path.isfile(timetagFilename):
    timetag = readFiletoString(timetagFilename)
# Read/generate buildTag...
if version and timetag:
    buildTag = 'v.' + version + '-' + timetag
elif path.isfile(buildTagFilename):
    buildTag = readFiletoString(buildTagFilename)

config = {  # Default config

    # Application parameters...

    'flaskEnv': flaskEnv,
    'gunicornEnv': gunicornEnv,
    'isGunicorn': isGunicorn,
    'isRPi': isRPi,
    'isDev': isDev,
    'isProd': isProd,
    'isTest': isTest,

    'pythonVersion': pythonVersion,
    'version': version,
    'timestamp': timestamp,
    'timetag': timetag,
    'buildTag': buildTag,

    'errorSendCode': True,
    'errorResponseType': 'text',
    #  'errorResponseType': 'json',

    # Path parameters...

    'rootPath': rootPath,  # RO!
    'uploadPath': uploadPath,

    # Generated client path (see `cam-client-app-build`, TODO?)

    #  'clientTemplatePath': path.join(rootPath, 'cam-client-app-build'),
    'clientTemplatePath': path.join(rootPath, 'src/templates'),
    'clientStaticPath': path.join(rootPath, 'static'),
    #  'clientStaticPath': path.join(clientTemplatePath, 'static'),
    'clientStaticUrl': '/static',
    #  'clientStaticUrl': '',

    # Logging...

    'outputLog': True,  # Print log to stdout
    'outputColoredLog': True,  # Use rich output log format with `termcolor`
    'writeLog': True,  # Write log to external file
    'clearLogFile': True,  # Clear log file at start
    'logFileName': 'log.txt',  # Log file name (relative to `rootPath`!)

    # Datetime formats...

    'dateTagFormat': '%y%m%d-%H%M',  # eg: '220208-0204'
    'dateTagPreciseFormat': '%y%m%d-%H%M%S',  # eg: '220208-020423'
    'shortDateFormat': '%Y.%m.%d %H:%M',  # eg: '2022.02.08-02:04'
    'preciseDateFormat': '%Y.%m.%d %H:%M:%S',  # eg: '2022.02.08-02:04:23'
    'logDateFormat': '%y%m%d-%H%M%S-%f',  # eg: '220208-020423-255157'
    'detailedDateFormat': '%Y.%m.%d %H:%M:%S.%f',  # eg: '2022.02.08-02:04:23.255157'

    # Databases...

    'dbPath': path.join(rootPath, 'db'),
    #  'dbExt': '.db',  # sqlite3 database file extensions
    'dbExt': '.json',  # tinydb database file extensions

    # API

    'legalOrigins': [
        # Real endpoint addresses...
        'https://tournament-game-build.march.team',
        'http://tournament-game-build.march.team',
    ],

    'apiRoot': '/api/v1.0',  # Eg: http://localhost:5000/api/v1.0/start
    'apiUser': 'api',
    'apiPass': 'pusplndvqaivbynv',  # Authorization: Basic YXBpOnB1c3BsbmR2cWFpdmJ5bnY=

    # Send mail

    'mailFromAddr': '"Tournament Site" <tournament@march.team>',
    'mailToAddr': ['tournament@march.team'],
    'mailToBCCAddr': ['dmia@yandex.ru'],

    'mailUser': 'site@march.team',
    'mailPass': 'M1qcqujpwflapqfgh',  # App pwd for `site@march.team`
    #  'mailUser': 'march.team.realty@gmail.com',
    #  'mailPass': 'mqcqujpwflapqfgh',  # App pwd for `march.team.realty@gmail.com`

}

# Allows CORS requests from developer (localhost) server
allowDebugOrigins = True

if allowDebugOrigins:
    config['legalOrigins'].append('http://localhost:3000')
    config['legalOrigins'].append('http://localhost:5000')


updateConfigWithYaml(config, yamlConfigFilename)
updateConfigWithYaml(config, yamlLocalConfigFilename)

#  #  DEBUG
#  print('Config:', config)
#  print('Done')
